File: examples/notifier/go/notifier.go
package notifier

import "fmt"

type Notifier interface {
	Send(message string)
}

type ConsoleNotifier struct{}

func (c ConsoleNotifier) Send(message string) {
	fmt.Printf("sent: %s\n", message)
}

type InMemoryNotifier struct {
	SentMessages []string
}

func (n *InMemoryNotifier) Send(message string) {
	n.SentMessages = append(n.SentMessages, message)
}

func SendReceiptReady(n Notifier, orderID string) {
	n.Send(fmt.Sprintf("Order %s is ready.", orderID))
}

File: examples/notifier/python/notifier/notifier.py
from typing import Protocol


class Notifier(Protocol):
    def send(self, message: str) -> None: ...


class ConsoleNotifier:
    def send(self, message: str) -> None:
        print(f"sent: {message}")


class InMemoryNotifier:
    def __init__(self) -> None:
        self.sent_messages: list[str] = []

    def send(self, message: str) -> None:
        self.sent_messages.append(message)


def send_receipt_ready(notifier: Notifier, order_id: str) -> None:
    notifier.send(f"Order {order_id} is ready.")
